### ISSUE TITLE
Config flow for HomeAssistant integration

### ISSUE THREAD

**Issue by farmio** [nielstron/pyfronius#40]
When trying to add a config flow for the HomeAssistant Fronius integration I encountered some challenges (some more than I'd expected 🤣). This issue is intended to keep track of these and discuss how to proceed.

- [ ] Find a unique_id for the `ConfigEntry`
Due to Fronius' unconsistent API usage its hard to find a unique_id for a config entry that fits API V0 and different API V1 (Snap-In and Gen24) devices. Also the reported `UniqueID` from eg. `GetInverterInfo` differs to `SerialNumber` from `GetActiveDeviceInfo.cgi?DeviceClass=System`. 
I'd like to get a unique_id for the device that we actually communicate with (the configured IP address), but IP addresses aren't valid unique_ids for HA. No endpoint reports mac-addresses, so I think the cleanest solution is to get the fronius `UniqueID` from `GetLoggerInfo`, and if not available (Gen24) fall back to `UniqueID` of smallest device_index (0 or 1, not sure right now) from `GetInverterInfo`.

- [ ] Look up available devices / endpoints
`GetActiveDeviceInfo.cgi\?DeviceClass\=System` is not available on API V0 so I guess we could query all supported system-scoped endpoints (inverter, storage, meter) to find out which devices are connected to the SolarNet and store this information in the ConfigEntry. So we can omit queries for non-existing devices later on (no Battery, no `GetStorageRealtimeData` requests after setup).

- [ ] Find a migration path
To migrate existing yaml configs to ConfigEntry we could just run the flow. This would not omit endpoints the user deliberately didn't configure in yaml. Also my plan was to only use system-scope queries from a flow-created setup. The easiest way would be to don't provide a migration path and let a user set up a config flow manually - we can't remove the yaml-config anyway so user action is always required (although we can ignore it when a ConfigEntry exists).

- [ ] Gracefully handle query errors
To be able to gracefully handle errors I think it would be nice if pyfronius raised just one error (or subclasses of a `PyfroniusError` that can be excepted). Right now it is `ConnectionError | ValueError | NotSupportedError` or just returning an empty dict.

- [ ] Typesafe return values
Currently returned value types are generic dicts not consistent with API keys. Eg. `GetActiveDeviceInfo`: `dict[str, dict[str, int | str] | list[dict[str, str]]]]` which means to get to an inverters serial number in a typesafe way we had to 
```python
if inverters := result.get("inverters") is not None:
    for inverter in inverters:
        if uid := inverter.get("UniqueID") is not None:
            return uid
```
using dataclasses or namedtuples could provide an easier interface for this, but would also mean the existing yaml-config logic would have to be adapted (which is not a huge effort, I suppose, but I currently struggle understanding how it works exactly).

Let me know your thoughts about any of these points 😀

**Comment by nielstron** [comment]
Below I will add some suggestions for the mentioned points

> Find a unique_id for the ConfigEntry

Provide the config flow only for devices that support a unique id. Otherwise, recommend a manual installation.

> Look up available devices / endpoints

We can provide an additional abstracted method in this package that returns a list of installed devices by first trying to fetch active devices, if the call is not available manually assemble information from the system scope calls. We should not be concerned with number of calls in the beginning, lets just see how user experience is working out.

> Gracefully handle query errors

FroniusError superclass sounds good to me. Note that an empty returned dict is not an error. It is just due to the fact that the fronius device did not return anything...

> Typesafe return values

Again, I would prefer to supply the user with an abstracted additional method within the pyfronius package, i.e. "unique_id(self)", which would automatically try to deduce a unique id for the device. Then the user would not have to worry about the setup.

**Comment by nielstron** [comment]
FroniusError is added as of https://github.com/nielstron/pyfronius/commit/6b3be4eb7da37c3703467fd902d264ff08b0f1d4

**Comment by farmio** [comment]
> FroniusError superclass sounds good to me. Note that an empty returned dict is not an error. It is just due to the fact that the fronius device did not return anything...

@nielstron I'd like to add raising an error when the API returned a status code != 0 (see https://github.com/farmio/pyfronius/commit/e2fa816e3d1bf66a9aeecb868c2cf17d275b0c7e)
Unfortunately the `fetch()` function and its tests currently rely on returning a dict with only the header.
Also there are some hints to break when errors occur, but it only loggs and passes currently. https://github.com/farmio/pyfronius/blob/e2fa816e3d1bf66a9aeecb868c2cf17d275b0c7e/pyfronius/__init__.py#L328 
Would it be safe to change the bahaviour of `fetch()` to suppress `BadStatusError` and not return anything for unsuccessful requests?
The intention of this is to guarantee to have a valid response from any request by just 
```python
try:
    response = await fronius.current_system_inverter_data()
except FroniusError:
    return
# use response
```
without having to check the status separately.

**Comment by farmio** [comment]
from https://github.com/home-assistant/core/blob/523998f8a11353fa360a698179eb773efd59f93b/homeassistant/components/fronius/sensor.py#L234
> Add discovered value fields as sensors
> because some fields are only sent temporarily

Sensor entities are created not only at start, but at runtime when new keys in the returned json appear. I couldn't find any hint in the API documentation regarding this behaviour. @nielstron do you think it is sufficient to create sensor entities from fields seen in the first successful response in `async_setup_entry`(previously `async_setup_platform`)?

**Comment by nielstron** [comment]
API V1 describes that in Meter and Storage requests, data might not be sent when devices are (temporarily, i.e. at night) offline. This matches my experience with my own device. I would therefore recommend detecting new values on the fly.

Regarding detection of devices it seems that active devices within the last 24 hours are usually listed, so this should suffice to query once (the user must reconfigure the component in case he/she installed new hardware, which sounds reasonable to me).

**Comment by nielstron** [comment]
To be more precise: Offline devices are still listed in meter and storage request, but information about the current power flow or similar values may be missing (the values are simply omitted in the response) if the device is not currently used.

**Comment by farmio** [comment]
Alright, found it! Thanks!
> Inactive channels are not included in the response and may vary depended on battery used and software version. Take care about permanently or temporary missing channels when processing this response.

Fronius doesn't make it especially straightforward 🤣 just having a `null` value would make things a lot easier.

**Comment by farmio** [comment]
So here is a first draft, if you want to have a look.
https://github.com/farmio/home-assistant/tree/fronius-refactor/homeassistant/components/fronius
Currently only meter devices are implemented. In theory multiple devices should work, but I have only one installed to test. Also runtime entity creation for previously missing keys should work, but also untested as my meter always returns the same json.

**Comment by nielstron** [comment]
I have almost zero experince/knowledge about config flows and automatic setup, so I cannot comment on code quality. Will try to run it once I get close to a fronius again :)

**Comment by farmio** [comment]
Storages should work too now. My knowledge about config flows was about the same 🤣 Automatic setup works by just using system scope endpoints for meters and storages. I guess for inverters, where we plan to continue using device scope, a reload of the integration would be necessary. But one doesn't add inverters that often I guess.

Don't you have a Fronius device anymore?

**Comment by nielstron** [comment]
> I guess for inverters, where we plan to continue using device scope, a reload of the integration would be necessary. But one doesn't add inverters that often I guess.

This sounds reasonable. So you would just do one manual system scope call and then create a bunch of data coordinators for each device?

**Comment by farmio** [comment]
Yes, pretty much. I'd like to do a `GetInverterInfo` request on setup. This returns solar_net_ids of every known inverter and also their unique_id and name. For every found inverter I'd set up a device scope coordinator. 

We could do a inverter System scope coordinator too to get cumulative data and just discard inverter data of this request - but when we see a unknown solar_net_id signal somehow to run inverter_info again and setup the new inverter device scope coordinator. 
... but that's quite much hassle for a very unlikely scenario of an added inverter. I'd postpone this for a later upgrade of the integration to keep it as simple as possible for the initial review.

**Comment by farmio** [comment]
Inverter device scope data is now implemented as well. It currently needs the pyfronius/dev branch to work (with #45 to get the model names).

**Comment by nielstron** [comment]
Thanks for pushing this forward! Looks all very week thought-through, hope to see the PR in HA go through well 😅

**Comment by farmio** [comment]
Thank you for your very quick support with all of my issues 😃
The PR to HA/core is now created https://github.com/home-assistant/core/pull/56166 - unfortunately it grew to a 1,4k lines diff 😱. Lets hope a HA core developer reviews it despite that.
